docs/utils/masked/index.rst

```diff
@@ -0,0 +1,165 @@
+masked |Quantity|.  For instance, consider an array ``s`` of lengths with
+shape ``(N, 3)``, in which the last axis represents width, height, and depth.
+With this, you could compute corresponding volumes by taking the product of
+the values in the last axis, ``s.prod(axis=-1)``. But if masked elements were
+skipped, the physical dimension of entries in the result would depend how many
+elements were masked, which is something |Quantity| could not represent (and
+would be rather surprising!).  As noted above, however, masked elements are
+skipped for operations for which this is well defined, such as for getting the
+mean and other sample properties such as the variance and standard deviation.
+
+A third difference is more conceptual.  For `~numpy.ma.MaskedArray`, the
+instance that is created is a masked version of the unmasked instance, i.e.,
+`~numpy.ma.MaskedArray` remembers that is has wrapped a subclass like
+|Quantity|, but does not share any of its methods.  Hence, even though the
+resulting class looks reasonable at first glance, it does not work as expected::
+
+  >>> q = [1., 2.] * u.m
+  >>> np_mq = np.ma.MaskedArray(q, mask=[False, True])
+  >>> np_mq
+  masked_Quantity(data=[1.0, --],
+                  mask=[False,  True],
+            fill_value=1e+20)
+  >>> np_mq.unit
+  Traceback (most recent call last):
+  ...
+  AttributeError: 'MaskedArray' object has no attribute 'unit'
+  >>> np_mq / u.s
+  <Quantity [1., 2.] 1 / s>
+
+In contrast, |Masked| is always wrapped around the data properly, i.e., a
+``MaskedQuantity`` is a quantity which has masked values, but with a unit that
+is never masked.  Indeed, one can see this from the class hierarchy::
+
+  >>> mq.__class__.__mro__
+  (<class 'astropy.utils.masked.core.MaskedQuantity'>,
+   <class 'astropy.units.quantity.Quantity'>,
+   <class 'astropy.utils.masked.core.MaskedNDArray'>,
+   <class 'astropy.utils.masked.core.Masked'>,
+   <class 'astropy.utils.shapes.NDArrayShapeMethods'>,
+   <class 'numpy.ndarray'>,
+   <class 'object'>)
+
+This choice has made the implementation much simpler: |Masked| only has to
+worry about how to deal with masked values, while |Quantity| can worry just
+about unit propagation, etc.  Indeed, an experiment showed that applying
+|Masked| to `~astropy.table.Column` (which is a subclass of `~numpy.ndarray`),
+the result is a new ``MaskedColumn`` that "just works", with no need for the
+overrides and special-casing that were needed to make `~numpy.ma.MaskedArray`
+work with `~astropy.table.Column`.  (Because the behaviour does change
+somewhat, however, we chose not to replace the existing implementation.)
+
+In some respects, rather than think of |Masked| as similar to
+`~numpy.ma.MaskedArray`, it may be more useful to think of |Masked| as similar
+to marking bad elements in arrays with NaN (not-a-number).  Like those NaN,
+the mask just propagates, except that for some operations like taking the mean
+the equivalence of `~numpy.nanmean` is used.
+
```

Comment:
Hmm, that example is I fear an immediate oversight, since
```
ma.min()
# 2
```
and surely `argmin()` and `min()` should be consistent!

And, yes, it is a really silly bug: I actually do mean to ignore all masked elements and to do that replace those with the maximum in the array, which I thought was good enough, but in this particular example, it ends up just doing `np.array([2, 2]).argmin()` which yields 0. Oops...

Bottom line is that clearly I should have the list you mentioned!

Reply:
On providing analogues to the `nan*` functions: I actually considered just overriding those functions to act on `ma.unmask(fill_value=np.nan)` (i.e., take masked elements to be `nan`, which needs some thought for integer dtype, obviously...); indeed, I have an extra commit that just does that but without any tests.

Reply:
I thought this infrastructure was meant to work on anything that could be in an array, e.g. strings or bool or whatever.

Reply:
Indeed, it should!  Which is probably one reason I didn't continue with the `nan*` overrides...

p.s. Fix for `argmin()`, `argmax()` coming up... Right now, like for `np.ma.MaskedArray`, I have it return `0` if all elements are masked. Should be a masked `0` instead?

Reply:
p.p.s. While it should be able to handle all dtype, I should be clear that the tests are almost all for float - plus a few for structured arrays consisting of floats - I re-use many for `Quantity` and `Longitude`...

Reply:
> Fix for argmin(), argmax() coming up... Right now, like for np.ma.MaskedArray, I have it return 0 if all elements are masked. Should be a masked 0 instead?

I think that case should be an exception like `np.argmin([])`. Returning 0 is definitely the wrong answer.
```
In [1]: np.argmin([])
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: attempt to get argmin of an empty sequence
```

Reply:
About the `nan*` functions, having analogs for those would be useful I think and also provide a bridge for migrating from `np.ma.MaskedArray`. They can certainly fail for non-applicable types (e.g. str) just as numpy now does. `np.nanmin(['a', 'b'])` gives a `TypeError: cannot perform reduce with flexible type`.

I guess this is not a hard requirement for this PR but I would strongly suggest as a followup.

Reply:
For the `nan*` functions, I was concretely suggesting to *override* those to mean any masked or nan elements are skipped. This as I don't really see those functions being very useful otherwise in the presence of masked values. But given the length of this PR and the thought required about how to implement it for integers, etc., perhaps best done as follow-up.

Reply:
@taldcroft - I realize we never resolved what `argmin/max` should do for the case that everything is masked. You suggested raising an error, but one problem with that is what should happen if one does `np.argmin(ma, axis=0)` and only some slices along axis 0 are fully masked? To me, treating it a bit more as "sort all values, with masked at the end, and then return the index of the value that is now at 0" is perhaps the best after all. 

But in principle, one could also return a masked index array.

Reply:
p.s. Added this to tracking issue #11529

Reply:
p.p.s. Also added the nanfunctions to #11529

Reply:
> what should happen if one does np.argmin(ma, axis=0) and only some slices along axis 0 are fully masked? 

I don't really see the `axis=0` case as being any different. The values you get out should be exactly the same if you do a Python loop over one axis of a 2-d array and compute argmin for each 1-d subset vs. using a single vectorized call.

Masking the value does not help in the current implementation, and in fact seems to allow silently getting the wrong answer:
```
In [12]: idx = Masked(0, mask=True)  
In [13]: idx
Out[13]: MaskedNDArray(—)
In [14]: np.arange(5)[idx]
Out[14]: 0  # Yikes, I expected an exception!
```

> "sort all values, with masked at the end, and then return the index of the value that is now at 0"

That seems like a weird ad hoc definition that fits the current code behavior, but nobody would really expect that. If you have no (valid) values in an array and ask for the index of the minimum, I cannot see any outcome besides an exception, or maybe a masked scalar, being reasonable. 

Returning a masked scalar could be acceptable as long as that value produces an exception if you try to index with it. I would set the data under the mask to be something not otherwise possible, for instance the MAXINT. In practice this probably won't be helping the user much because it will move the exception (from subsequent indexing with a masked scalar) away from the actual problem. But when you start thinking about how to make up arbitrary values, that is a sign of a problem.

Reply:
Unfortunately, I cannot control what `np.ndarray` does with a masked index...  Elsewhere, I just decided not to allow masked indices at all, since it seems very unclear what should happen. But that then suggests the choice is between returning some number or raising an exception on any fully-masked slice. It is tricky... But let's move discussion to #11539...